sync/handlers/leafs_request_test.go

```diff
@@ -649,6 +651,51 @@ func TestLeafsRequestHandler_OnLeafsRequest(t *testing.T) {
 				assertRangeProofIsValid(t, &request, &leafsResponse, false)
 			},
 		},
+		"reading from snapshot with stale data times out": {
+			prepareTestFn: func() (context.Context, message.LeafsRequest) {
+				// delayedReader allows us to cause the snapshot read context
+				// to timeout, so we can test the trie is read from disk as a fallback.
+				db := &delayedReader{
+					KeyValueStore: memdb,
+				}
+				snap, err := snapshot.New(db, trieDB, 64, common.Hash{}, accountTrieRoot, false, true, false)
+				if err != nil {
+					t.Fatal(err)
+				}
+				// delete the first key from the snapshot so that we can simulate
+				// outdated snapshot data.
+				it := snap.DiskStorageIterator(smallStorageAccount, common.Hash{})
+				defer it.Release()
+				var firstKey common.Hash
+				for it.Next() {
+					firstKey = it.Hash()
+					break
+				}
+				rawdb.DeleteStorageSnapshot(memdb, smallStorageAccount, firstKey)
+
+				// set the delay after creating the snapshot to avoid slowing the test
+				db.delay = 10 * time.Millisecond
+				snapshotProvider.Snapshot = snap
+
+				ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
+				t.Cleanup(cancel) // avoids leaking the context
+
+				return ctx, message.LeafsRequest{
+					Root:    smallTrieRoot,
+					Account: smallStorageAccount,
+					Limit:   maxLeavesLimit,
+				}
+			},
+			assertResponseFn: func(t *testing.T, request message.LeafsRequest, response []byte, err error) {
+				require.NoError(t, err)
+				var leafsResponse message.LeafsResponse
+				_, err = message.Codec.Unmarshal(response, &leafsResponse)
+				require.NoError(t, err)
+				require.NotZero(t, mockHandlerStats.SnapshotReadAttemptCount)
+				require.Zero(t, mockHandlerStats.SnapshotReadSuccessCount)
+				require.NotEmpty(t, leafsResponse.Keys)
```

Comment:
Can we require that it returns the full 500 keys/values here instead of not empty?

I assume this was done to handle the case that serving each section from the trie times out, but this should pass consistently for the local test with an in-memory database.